Winton.Extensions.Threading.Actor/Internal/ActorTaskScheduler.cs

```diff
@@ -51,206 +53,261 @@ public ActorPauseAwaitable<T> WhileActorPaused<T>(Task<T> task)
-                        if (actorTaskContext.Kind == ActorTaskKind.Resumer)
+                case ActorTaskSchedulerStatus.Terminated:
+                    ReleaseLock();
+                    CancelActorTask(actorTask);
+                    break;
+                case ActorTaskSchedulerStatus.Inactive:
+                    _status = ActorTaskSchedulerStatus.Active;
+                    ReleaseLock();
+                    QueueForExecution(actorTask);
+                    break;
+                case ActorTaskSchedulerStatus.Active:
+                    if ((actorTask.Traits & ActorTaskTraits.Resuming) == ActorTaskTraits.Resuming)
+                    {
+                        actorTask.Next = _front;
+                        _front = actorTask;
+
+                        if (_back is null)
                         {
-                            _statusManager.MarkActive();
-
-                            if (HaveActiveThread)
-                            {
-                                _resumingTask = task;
-                            }
-                            else
-                            {
-                                LaunchNew(task);
-                            }
+                            _back = _front;
+                        }
+                    }
+                    else
+                    {
+                        if (_front is null)
+                        {
+                            _front = actorTask;
                         }
                         else
                         {
-                            _taskQueue.Enqueue(task);
+                            _back.Next = actorTask;
                         }
-                        break;
-                    default:
-                        throw new ArgumentOutOfRangeException();
-                }
+
+                        _back = actorTask;
+                    }
+
+                    ReleaseLock();
+
+                    break;
+                case ActorTaskSchedulerStatus.Paused:
+                    if ((actorTask.Traits & ActorTaskTraits.Resuming) == ActorTaskTraits.Resuming)
+                    {
+                        _status = ActorTaskSchedulerStatus.Active;
+                        ReleaseLock();
+                        QueueForExecution(actorTask);
+                    }
+                    else
+                    {
+                        if (_front is null)
+                        {
+                            _front = actorTask;
+                        }
+                        else
+                        {
+                            _back.Next = actorTask;
+                        }
+
+                        _back = actorTask;
+                        ReleaseLock();
+                    }
+
+                    break;
             }
         }
 
-        protected override bool TryExecuteTaskInline(Task task, bool taskWasPreviouslyQueued)
+        private void TakeLock()
```

Comment:
Note that the taking and releasing of the lock was isolated partly in order to experiment with using a `SpinLock`. However, using a `SpinLock` made no observable difference, so I've stuck with using a standard lock.